draft-js-inline-toolbar-plugin/src/components/Toolbar/index.js

```diff
@@ -45,23 +61,126 @@ export default class Toolbar extends React.Component {
+      const beforeParentEnd = relativeRect.right - fromBeginningToMiddle;
+
+      // the selection is closer to parent beginning than half of the toolbar
+      // +-----------------------------------------------+
+      // |          vv toolbar                           |
+      // | +------------------+                          |
+      // | +------------------+                          |
+      // |                                               |
+      // |  +--+                                         |
+      // |   ^^ selection                                |
+      // +-----------------------------------------------+
+      if (fromParentBeginning < toolbarHalfWidth) {
+        metrics.ruleValue = toolbarHalfWidth;
+        metrics.shift = 'left';
+      } else if (beforeParentEnd < toolbarHalfWidth) {
+        // the same, but relative to the parent end
+        // +-----------------------------------------------+
+        // |                                 vvv toolbar   |
+        // |                            +---------------+  |
+        // |                            +---------------+  |
+        // |                                               |
+        // |                                      +--+     |
+        // |                             selection ^^      |
+        // |                                               |
+        // +-----------------------------------------------+
+        metrics.ruleValue = -toolbarHalfWidth;
+        metrics.rule = 'right';
+        metrics.shift = 'right';
+      } else {
+        // selection somewhere in the middle within the parent and there is a
+        // free place for toolbar
+        metrics.ruleValue = (selectionRect.left - relativeRect.left)
+          + (selectionRect.width / 2);
+      }
+
+      // ok, why we have to use external style elements here? Because, we are
+      // moving pointer position each time and it can be set via `before`,
+      // `after` css-pseudo classes
+      const pointerClassName = 'draft-js-inline-toolbar-pointer';
+      const styleId = 'draft-js-inline-toolbar-pointer-pseudo';
+
+      if (this.toolbar.classList.contains(pointerClassName)) {
+        this.toolbar.classList.remove(pointerClassName);
+      }
+      // We are recreating style element on every toolbar appearance. It is
+      // better for us to use individual style element rather than join new
+      // styles to some of the existent elements, because we can easily remove
+      // it next time
+      if (document.getElementById(`${styleId}-${this.state.nonce}`)) {
+        document.head.removeChild(
+          document.getElementById(`${styleId}-${this.state.nonce}`)
+        );
+      }
+
+      if (typeof metrics.shift === 'string') {
+        const styleEl = document.createElement('style');
+        styleEl.setAttribute('id', `${styleId}-${this.state.nonce}`);
+        document.head.appendChild(styleEl);
+        this.toolbar.classList.add(pointerClassName);
```

Comment:
this is a big no-no in react land. One should only use the react api to affect dom-changes, otherwise the virtual DOM and the actual DOM diverge and that almost certainly will cause exceptions or bugs.

Reply:
Ok, this is the most unclear part to me. Do you know how we can change pointer position, read "css pseudo-class properties" with react / inline styling?

Reply:
Hummm. Good point. I like what you did there, it's quite nicely done :+1:. Let me have a tinker with this. I might have time today or tomorrow morning :+1: